File: src/crypto/zkp/pedersen_vector_commitment.py
import gmpy2 as gmpy

class pedersen_vector_commitment:
    def __init__(self, p, q, g, h):
        self.p = gmpy.mpz(p)
        self.q = gmpy.mpz(q)
        temp = []
        for i in range(len(g)):
            temp.append(gmpy.mpz(g[i]))
        self.g = temp
        self.h = gmpy.mpz(h)
    
    def generateCommitment(self, vector, r):
        if len(vector) != len(self.g):
            raise ValueError("Size of vector does not match")

        c = gmpy.powmod(self.h, r, self.p)
        for i in range(len(self.g)):
            if(gmpy.mpz(vector[i]) < 0):
                inverse = gmpy.invert(self.g[i], self.p)
                c = gmpy.mul(c, gmpy.powmod(inverse,gmpy.mul(gmpy.mpz(-1), vector[i]), self.p))
            else :
                c = gmpy.mul(c, gmpy.powmod(self.g[i], vector[i], self.p))
            
        c = c % self.p
        return c

    #def getRandomZq():
        #bits = self.q.bitLength()
        #while(True):
            #r = gmpy.mpz(random(help.getRandomInt(2, bits)));
            #if(r.compare(self.q) == -1) return r;

    def get_setup_values(self):
        p = self.p
        q = self.q
        g = self.g
        h = self.h
        return p, q, g, h      
 
    def get_p(self) :
        return self.p

    def get_q(self) :
        return self.q    

    def get_g(self) :
        return self.g

    def get_h(self) :
        return self.h
        

File: gui/app/src/components/Verification/ZKP/PedersenVectorCommitment.js
import help from "src/components/HelpFunctions";
import random from 'crypto-random-bigint';
var bigInt = require("big-integer");

export class PedersenVectorCommitment {
    /**
     * Initilazies a PedersenVectorCommitment instance
     * @param {String} p prime p with p = 2q + 1
     * @param {String} q prime q with p = 2q + 1
     * @param {Array.<String>} g generators of cyclic group with order q
     * @param {String} h h
     */
    constructor(p, q, g, h) {
        this.p = bigInt(p);
        this.q = bigInt(q);
        var temp = new Array(g.length);
        for(var i = 0; i < g.length; i++) {
            temp[i] = bigInt(g[i]);
        }
        this.g = temp;
        this.h = bigInt(h);
    }

    /**
     * computes the commitment of a vector
     * @param {Array.<bigInt>} vector values for which the commitment is computet
     * @param {bigInt} r random number from Z_q used in the commitment
     * @returns {bigInt} commitment
     */
    generateCommitment(vector, r) {
        if(vector.length != this.g.length) {
            throw new Error("Size of vector does not match");
        }

        //TODO: make sure mod p ist right
        var c = this.h.modPow(r, this.p);
        for(var i = 0; i < this.g.length; i++) {
            if(bigInt(vector[i]).isNegative()) {
                var inverse = this.g[i].modInv(this.p);
                c = c.multiply(inverse.modPow(bigInt(-1).multiply(vector[i]), this.p));
            } else {
                c = c.multiply(this.g[i].modPow(vector[i], this.p));
            }
        }
        c = c.mod(this.p); //TODO: make sure mod p ist right, also if it is use mod p in all operations
    
        return c;
    }

    /**
     * gets a random elment from Z_q
     * @returns random element from Z_q
     */
    getRandomZq(){
        var bits = this.q.bitLength();
        while(true){
            var r = bigInt(random(help.getRandomInt(2, bits)));
            if(r.compare(this.q) == -1) return r;
        }
    }

    /**
     * gets the setup values of the commitment scheme
     * @returns setup values of the commitment scheme //TODO: return type
     */
    getSetupValues() {
        var p = this.p;
        var q = this.q;
        var g = this.g;
        var h = this.h;
        return  {p, q, g, h};
    }

    /**
     * gets prime p
     * @returns {bigInt} prime p with p = 2q + 1
     */
    getP() {
        return this.p;
    }

    /**
     * gets prime q
     * @returns {bigInt} prime q with p = 2q + 1
     */
    getQ() {
        return this.q;
    }

    /**
     * gets the generators used in the commitment scheme
     * @returns {Array.<bigInt>} generators
     */
    getG() {
        return this.g;
    }

    /**
     * gets h
     * @returns {bigInt} h
     */
    getH() {
        return this.h;
    }
}